examples/scann.py

```diff
@@ -0,0 +1,261 @@
+)
+# Train-test split.
+train_ratings = shuffled_ratings.take(80_000).batch(1000).cache()
+test_ratings = shuffled_ratings.skip(80_000).take(20_000).batch(1000).cache()
+
+"""
+## Implementing the Model
+"""
+
+
+class RetrievalModel(keras.Model):
+    def __init__(
+        self,
+        num_users,
+        num_candidates,
+        embedding_dimension=32,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        # Our query tower, simply an embedding table.
+        self.user_embedding = keras.layers.Embedding(
+            num_users, embedding_dimension
+        )
+        # Our candidate tower, simply an embedding table.
+        self.candidate_embedding = keras.layers.Embedding(
+            num_candidates, embedding_dimension
+        )
+
+        self.loss_fn = keras.losses.MeanSquaredError()
+
+    def build(self, input_shape):
+        self.user_embedding.build(input_shape)
+        self.candidate_embedding.build(input_shape)
+
+        super().build(input_shape)
+
+    def call(self, inputs, training=False):
+        user_embeddings = self.user_embedding(inputs)
+        result = {
+            "user_embeddings": user_embeddings,
+        }
+        return result
+
+    def compute_loss(self, x, y, y_pred, sample_weight, training=True):
+        candidate_id, rating = y["movie_id"], y["rating"]
+        user_embeddings = y_pred["user_embeddings"]
+        candidate_embeddings = self.candidate_embedding(candidate_id)
+
+        labels = keras.ops.expand_dims(rating, -1)
+        # Compute the affinity score by multiplying the two embeddings.
+        scores = keras.ops.sum(
+            keras.ops.multiply(user_embeddings, candidate_embeddings),
+            axis=1,
+            keepdims=True,
+        )
+        return self.loss_fn(labels, scores, sample_weight)
+
+
+"""
+## Training the model
+"""
+
+model = RetrievalModel(users_count + 1000, movies_count + 1000)
+model.compile(optimizer=keras.optimizers.Adagrad(learning_rate=0.1))
+
+history = model.fit(
+    train_ratings, validation_data=test_ratings, validation_freq=5, epochs=50
+)
+
+"""
+## Making predictions
+
+Before we try out ScANN, let's go with the brute force method, i.e., for a given
+user, scores are computed for all movies, sorted and then the top-k
+movies are picked. This is, of course, not very scalable when we have a huge
+number of movies.
+"""
+
+candidate_embeddings = keras.ops.array(
+    model.candidate_embedding.embeddings.numpy()
+)
+# Artificially duplicate candidate embeddings to simulate a large number of
+# movies.
+candidate_embeddings = keras.ops.concatenate(
+    [candidate_embeddings]
+    + [
+        candidate_embeddings
+        * keras.random.uniform(keras.ops.shape(candidate_embeddings))
+        for _ in range(100)
+    ],
+    axis=0,
+)
+
+user_embedding = model.user_embedding(keras.ops.array([10, 5, 42, 345]))
+
+print(f"{candidate_embeddings.shape=}", f"{user_embedding.shape=}")
```

Comment:
Is it intentional that this `print` is here?

Reply:
Oh, sorry. I was using it for debugging